### DISCUSSION TITLE
Target-typed object initializers

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#5671] [Language Ideas]
Based on a discussion around extension methods in object initializers the other day, I had an idea about how we could use [roles](https://github.com/dotnet/csharplang/issues/5497) to form the basis of DSL framework for C#, with object creation and object initializers as the backbone of the feature. This is still somewhat rough (and in particular, I haven't had super deep thoughts about how to actually make the target-typing work beyond "eh, we've done harder things, we can probably make it work"), but I figured that the community might like to read it and have other ideas as to where we could take this, so here it is.

# Target-typed object initializers

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Started

## Summary
[summary]: #summary

C# object initializers today work very well for types that know their set of properties when the type is declared. This is fine for most scenarios, but a few scenarios, particularly in UI layouts, don't work this way. We could improve these scenarios with a combination of the roles proposal and allowing extension properties/fields to be set in object initializers when an object is being converted to a role that adds such properties. To a lesser extent, this could also work with explicitly-implemented properties from interfaces, but that is likely not extensible enough for a general library to be based around this.

## Motivation
[motivation]: #motivation

We've had a number of requests over the years for the ability to build DSLs in C#, often inspired by other languages such as Kotlin or Swift with their trailing delegate syntax. These languages use trailing delegate syntax to allow for deeply-nested lambdas, and have build a few DSLs around this feature, including Jetpack Compose and SwiftUI:

```kotlin
// Jetpack Compose layout creating a button with an icon and some text
Button(
    onClick = { /* ... */ },
    contentPadding = PaddingValues(start = 20.dp, top = 12.dp, end = 20.dp, bottom = 12.dp)
) {
    // This inner lambda implicitly changes what `this` refers to, so `Icon` is
    // actually being called on the containing `Button` instance
    Text("Like")
    Icon(Icons.Filled.Favorite)
}
```

```swift
// SwiftUI equivalent
Button(action: { /* ... */ }) {
    Label("Like", systemImage: "Icons.Filled.Favorite")
}.padding(EdgeInsets(top: 12, leading: 20, bottom: 12, trailing: 20))
```

For these simple scenarios, C# object initializers work well today. Some pseudo-syntax equivalents:

```cs
new Button(onClick: () => { /* ... */})
{
    Content =
    {
        new Label { Text = "Like" },
        new Icon { Path = "Icons.Filled.Favorite" }
    },
    Padding = new Padding
    {
        Start = 20,
        End = 20,
        Top = 12,
        Bottom = 12
    }
}
```

This is a reasonable approximation of the former examples, and already works well in C# today. This will become even more fluent and consistent with https://github.com/dotnet/csharplang/issues/5176, which will allow the `onClick` handler to be moved to the object initializer as well. Where C# starts to fall down, however, is setting _modifiers_ on these elements, based on the containing type. For example, imagine there's a grid containing the above button. In order to inform the containing grid what row/column this button should go in, a few strategies can be employed:

1. Make a `GridItem` containing type that has `Row` and `Column` properties. This allows use of standard object initialization syntax, but inserts an extra object in the tree that is purely there as a container, which is not particularly ergonomic.
2. Implement extension methods for setting the `Row` and `Column` properties, which are implemented as dependency properties or a property bag. This isn't great either: these extensions aren't usable in the object initialization syntax, and that makes for an inconsistent experience. These extensions can also be called on _any_ control, rather than just on the controls that are nested inside a `Grid`.

This is where target-typed object initializers and roles can come in and help. The following example makes use of a role for all grid children that adds a Row and Column extension property. The compiler then infers that those properties should be available in the object initializer based on the target type of the object creation expression, allowing the user to set them directly. I'm also making use of the collection literal syntax proposed by https://github.com/dotnet/csharplang/issues/5354 to create the child list.

```cs
public class UIControl { ... } // Base type for all controls

public class Grid : UIControl
{
    public IEnumerable<GridChildControl> Children { get; init; }
}

public role GridChildControl : UIControl
{
    // Implementations elided
    public int Row { get; init; }
    public int Column { get; init; }
}

Grid g = new Grid()
{
    NumRows = 1,
    NumColumns = 1,
    Children = [
        new Label
        {
            Text = "Hello World!",
            Row = 0, // Because `Children` is an `IEnumerable<GridChildControl>`, `Row` and `Column` are available to be set
            Column = 0
        }
    ]
};
```

This allows us to have a flexible, type-safe DSL for creating declarative UIs in C#. This can further enhanced with a few other proposals for further flexibility:

* https://github.com/dotnet/csharplang/discussions/4622 - Combined object and collection initializers. This would allow for children of the `Grid` to be directly added to the grid as well as being able to set properties on it at the same time, removing a nesting level by removing the `Children` property.
* https://github.com/dotnet/csharplang/issues/803 - Allow object initializers to be applied to the return of factory methods. Today, common helpers for creating UI elements would have to be new constructors, or take all info as parameters. If we allowed them to be methods instead, that would enhance the flexibility of the framework and allow for better customization.
* https://github.com/dotnet/csharplang/issues/3086 - Some form of expression block would bring a lot of power to this scenario, particularly if the expression block is able to `yield` elements to a containing collection initializer.
* https://github.com/dotnet/csharplang/issues/2556 - Similar to the previous suggestion, but perhaps more narrowly scoped to object initializers, this could add a lot of power to the syntax form.

## Detailed design
[design]: #detailed-design

To be filled out in more detail. For now, simply assume that we work some magic to be able to determine that, even for nested generic cases (as in the list example above), we can work out that the child elements will be subject to a conversion to another type, and that we can then use that information to add any members not currently visible to the set of available members.

In particular, this could allow explicitly-implemented members to become visible via their interface, if the target type was an interface. This could be used to implement a less-extensible version of the DSL, but this would require a number of extraneous interfaces and doesn't allow extending the feature to custom components (as you couldn't implement the extraneous interfaces on the original control types).

## Drawbacks
[drawbacks]: #drawbacks

All target-typed features add complication, and this one could be a doozy. In particular, it will have potential negative interactions with required members: what if an extension property needs a value that hasn't been set yet because the required member has not been initialized? Bringing explicitly-implemented members back into scope could also contravene the API author's design: they hid it for a reason. A forthcoming proposal on accessing a specific version of a virtual member might be a better approach for that part of the feature.

Building a DSL language on top of object creation is also a potential issue: caching unchanged layouts might not be possible, or might become more difficult, if this approach is taken. A method-based approach with trailing delegate syntax might be the preferred way to go in these cases.

## Alternatives
[alternatives]: #alternatives

Using a different approach entirely for DSL creation, such as the trailing delegate syntax that has been proposed a few times. This fits less well with our existing syntax forms (object initializers have been in C# for a while), but it could offer more flexibility in graph construction.

## Unresolved questions
[unresolved]: #unresolved-questions

TBD.

## Design meetings

Link to design notes that affect this proposal, and describe in one sentence for each what changes they led to.

**Comment by TahirAhmadov** [comment]
A question for those working with WPF/XAML and other similar frameworks: why is there an emphasis on using `Row`/`Column` type "properties" to assign the control's placement, instead of just using good old table-row-cell syntax similar to HTML?
```cs
new Grid
{
  new Row
  {
    new Button { ... },
    new BlankSpace(),
  },
  new Row
  {
    new BlankSpace(),
    new TextBox { ... },
  },
}
```

**Reply by theunrepentantgeek** [reply]
Your example @TahirAhmadov assumes each cell gets one thing - exactly one thing - in my experience, this is only true for very simple layouts.

Specifying Row, RowSpan, Column, and ColumnSpan individually for each control in order allows more flexibility:

* Tab order is defined by the order of control creation. Often tab order needs to diverge from a simple top-down, left-right (or right-left) model to give a good user experience. This is especially common if you're paying attention to accessibility.

* Overlapping controls are common. Two examples - sometimes a tinted panel will be put behind a set of controls to visually group them; and sometimes a button (or other control) will overlap another one in one corner.

* Collapsing rows (and sometimes columns) are a common technique to create dynamic and adaptive experiences. Common examples include error & validation messages that appear only when necessary, and administrative features that only show for users with sufficient privilege. 

Lastly, an observation. 

The table model of HTML was defined around 1995 (it wasn't a part of the initial release in 1993 but emerged pretty quickly). 

When the WPF folks were designing their object model in 2005 or so (WPF was released in late 2006) the industry had a decade of experience working with HTML tables. Don't you think they would have simply reused the older model if it were sufficient to the task at hand?

**Reply by MgSam** [reply]
I haven't tried MAUI or Blazor but I'm assuming this feature would hopefully be relevant for them? Would seem bizarre to add language features with the goal of making a dying 16 year old technology (WPF) slightly easier to write.

**Reply by TahirAhmadov** [reply]
@theunrepentantgeek all your examples are also possible with the "structured approach" I described; I showed the simplest scenario on purpose not to complicated the discussion. 
My take on it is that it's easier to specify tab index as a property of a control than set its position using pseudo properties. They had to invent a special syntax for it in XAML and now we're talking about adding special syntax to C#. At some point, I think we have to weigh the costs and benefits of this. 
Lastly, not everything new is better. By no means should HTML tables be used for layout, but a structured layout where you set properties on the objects themselves is more natural than this whole pseudo properties approach. 

@MgSam yes I know WPF is not the important framework here - the same approach is used in newer frameworks, too. Hence, my question is still relevant today. I wrote "and similar frameworks" to cover these newer ones.

**Reply by 333fred** [reply]
> I haven't tried MAUI or Blazor but I'm assuming this feature would hopefully be relevant for them? Would seem bizarre to add language features with the goal of making a dying 16 year old technology (WPF) slightly easier to write.

Indeed, my hope is that this is useful for _any_ C#-based DSL, not just WPF. Row and Column are just simple examples that popped to my mind immediately because the little UI work I did on Android in the past involved grids. UI frameworks often rely on these types of attached properties that should only exist in specific contexts, and my discussion with some of the comet folks the other day indicated that they're no different.

**Reply by HaloFour** [reply]
IIRC WPF attached properties aren't strictly limited to contexts like this and you can technically attach a `Grid.Row` value to any `UIControl` even if it's not the child of a `Grid`.  That might not impact the behavior of the control within the grid, but theoretically other code may reuse those properties as they see fit.  I think that there are a number of attached properties that could be attached to anything, like `ToolTipService`.

I guess those scenarios could be handled by extensions rather than roles, tho.

**Comment by saint4eva** [comment]
Comet UI MVU syntax would be more perfect for this.

**Reply by 333fred** [reply]
I actually talked about this general idea a bit with @Clancey before putting the proposal up. We think there's promise with the approach. One of the potential issues with Comet (and has been an issue with dependency properties in C# in general) is that we don't restrict where properties can be set, which can lead to unintuitive API patterns and an inability to know what things can be set in the current context.
